### PR TITLE
[ticket/16904] Fix regression in MCP for topics selection

### PR DESCRIPTION
**Pull request by rxu**
Regression was introduced in #5760.

Checklist:

- [x] Correct branch: master for new features; 3.3.x for fixes
- [x] Refactor MCP tests
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/master/coding-guidelines.html) and [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

<a href="https://tracker.phpbb.com/browse/PHPBB3-16904">PHPBB3-16904</a>.

### CODE REVIEW

**Comment by rxu** [comment]
Currently this PR also changes behavior of `function phpbb_extra_url()` so that its resulting string is not being prepended with `&amp`.

File: tests/functional/mcp/mcp_main_test.php [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+	 * @depends test_create_topics
+	 */
+	public function test_mcp_view_forum($post)
+	{
+		$this->add_lang(['common']);
+		$this->login();
+
+		// Browse MCP main page from forum view (gives &f=2)
+		$crawler = self::request('GET', "viewforum.php?f=2&sid={$this->sid}");
+		$mcp_link = substr_replace($crawler->selectLink($this->lang('MCP_SHORT'))->attr('href'), '', 0, 2); // Remove leading ./
```
Comment (marc1706) [inline tests/functional/mcp/mcp_main_test.php:57]: How about just using:
```
$crawler = self::$client->click($crawler->selectLink($this->lang('MCP_SHORT'))->link());
```

File: tests/functional/mcp/mcp_main_test.php [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+	public function test_mcp_view_forum_actions($action, $message, $require_confirmation, $post)
+	{
+		$topic_id_1 = $post[0]['topic_id'];
+		$topic_id_2 = $post[1]['topic_id'];
+
+		$this->add_lang(['common', 'mcp']);
+		$this->login();
+
+		$crawler = self::request('GET', "viewforum.php?f=2&sid={$this->sid}");
+		$mcp_link = substr_replace($crawler->selectLink($this->lang('MCP_SHORT'))->attr('href'), '', 0, 2); // Remove leading ./
```
Comment (marc1706) [inline tests/functional/mcp/mcp_main_test.php:98]: See above comment.

File: tests/functional/mcp/mcp_main_test.php [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+		{
+			if ($action == 'merge_topics')
+			{
+				// Merge topic_id_1 into topic_id_2
+				$select_for_merge_link = substr_replace($crawler->filter('.row a')->reduce(
+					function ($node, $i) use ($topic_id_2)
+					{
+						return (bool) strpos($node->attr('href'), "to_topic_id=$topic_id_2");
+					}
+				)->attr('href'), '', 0, 2); // Remove leading ./
```
Comment (marc1706) [inline tests/functional/mcp/mcp_main_test.php:118]: As stated above, there is a better way than trying to do substr_replace() ;)

File: tests/functional/mcp/mcp_test.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+class phpbb_functional_mcp_test extends phpbb_functional_test_case
+{
+	public function test_all_mcp_module_links()
+	{
+		$this->login();
+		$this->add_lang(['common', 'mcp']);
+
+		// Browse MCP main page
+		$crawler = self::request('GET', 'index.php');
+		$mcp_link = substr_replace($crawler->selectLink($this->lang('MCP_SHORT'))->attr('href'), '', 0, 2); // Remove leading ./
```
Comment (marc1706) [inline tests/functional/mcp/mcp_test.php:26]: Same as in the mcp_main_test

**Review by marc1706 — CHANGES_REQUESTED** [review]

File: tests/functional/mcp/mcp_main_test.php [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+
+				$crawler = self::request('GET', $select_for_merge_link);
+			}
+
+			$form = $crawler->selectButton($this->lang('YES'))->form();
+
+			if (in_array($action, ['fork', 'move']))
+			{
+				// Fork or move the topic to the forum id=3 'Download #1'
```
[the remark this answers is not in the document]
Comment (rxu) [reply tests/functional/mcp/mcp_main_test.php:127]: Shall I create a forum just for that?

**Review by CHItA — APPROVED** [review]

**Review by marc1706 — APPROVED** [review]

**Comment by marc1706** [comment]
@rxu could you create a PR to resolve the merge conflicts with master? right now it does not properly merge into master and I won't have time to have a look until beginning of next week.

**Comment by rxu** [comment]
@marc1706 Done.

**Pull request merged**